src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperParseTreeRewriter.cs

```diff
@@ -450,11 +504,23 @@ private void BuildCurrentlyTrackedBlock()
 
         private void BuildCurrentlyTrackedTagHelperBlock(Block endTag)
         {
+            Debug.Assert(_trackerStack.Any(tracker => tracker.IsTagHelper));
+
+            // We need to pop all trackers until we reach our TagHelperBlock. We can throw away any non-TagHelper
+            // trackers because they don't need to be well-formed.
+            TagHelperBlockTracker tagHelperTracker;
+            do
+            {
+                tagHelperTracker = PopTrackerStack() as TagHelperBlockTracker;
+            }
+            while (tagHelperTracker == null);
+
             // Track the original end tag so the editor knows where each piece of the TagHelperBlock lies
             // for formatting.
-            _trackerStack.Pop().Builder.SourceEndTag = endTag;
+            tagHelperTracker.Builder.SourceEndTag = endTag;
 
-            _currentTagHelperTracker = _trackerStack.Count > 0 ? _trackerStack.Peek() : null;
+            _currentTagHelperTracker =
+                _trackerStack.FirstOrDefault(tagBlockTracker => tagBlockTracker.IsTagHelper) as TagHelperBlockTracker;
```

Comment:
The `as` is there in-case there is no `TagHelper` block, aka `FirstOrDefault` returns `null`